src/shared/varlink.c

```diff
@@ -412,10 +412,10 @@ static int varlink_test_disconnect(Varlink *v) {
         if (IN_SET(v->state, VARLINK_AWAITING_REPLY, VARLINK_AWAITING_REPLY_MORE, VARLINK_CALLING, VARLINK_IDLE_SERVER) && v->read_disconnected)
                 goto disconnect;
 
-        /* Similar, if are a client that hasn't written anything yet but the write side is dead, also
+        /* Similar, if we are a client that has yet to finish writing or hasn't written anything yet, but the write side is dead, also
          * disconnect. We also explicitly check for POLLHUP here since we likely won't notice the write side
          * being down if we never wrote anything. */
-        if (IN_SET(v->state, VARLINK_IDLE_CLIENT) && (v->write_disconnected || v->got_pollhup))
+        if (IN_SET(v->state, VARLINK_PENDING_METHOD, VARLINK_PENDING_METHOD_MORE, VARLINK_IDLE_CLIENT) && (v->write_disconnected || v->got_pollhup))
```

Comment:
@poettering  Before this change, `systemd-oomd` sends a method call with `more` and can stop/disconnect, leaving pid1 in `VARLINK_PENDING_METHOD_MORE`. pid1 then continually gets POLLHUP on its fd. Since this case isn't handled, pid1 repeatedly calls `io_callback` as a result and causes the manager loop to throttle.

Since pid1 is the write side in this case, in a state where it is expected to write, but got a POLLHUP, shouldn't it disconnect?